Expansion of type_id fails if rename is used for cxx crate

```toml
[dependencies]
cxx_fork = { package = "cxx", version = "1.0" }
```

```rust
use cxx_fork::{type_id, ExternType};

unsafe impl ExternType for Foo {
    type Id = type_id!("Foo");
    type Kind = cxx_fork::kind::Opaque;
}
```

```console
error[E0433]: failed to resolve: could not find `cxx` in crate root
 --> src/main.rs:4:15
  |
4 |     type Id = type_id!("Foo");
  |               ^^^^^^^^^^^^^^^ could not find `cxx` in crate root
```

It should be possible to support this by making use of `$crate` inside of the macro.

--- a/macro/src/expand.rs
+++ b/macro/src/expand.rs
@@ -10,6 +10,7 @@ use crate::syntax::{
     self, check, mangle, Api, Doc, Enum, ExternFn, ExternType, Impl, Lifetimes, Pair, Signature,
     Struct, Trait, Type, TypeAlias, Types,
 };
+use crate::type_id::Crate;
 use proc_macro2::{Ident, Span, TokenStream};
 use quote::{format_ident, quote, quote_spanned, ToTokens};
 use std::mem;
@@ -1095,7 +1096,7 @@ fn type_id(name: &Pair) -> TokenStream {
     segments.extend(namespace_segments.cloned());
     segments.push(Ident::new(&name.cxx.to_string(), Span::call_site()));
     let qualified = QualifiedName { segments };
-    crate::type_id::expand(qualified)
+    crate::type_id::expand(Crate::Cxx, qualified)
 }
 
 fn expand_rust_box(ident: &Ident, types: &Types, explicit_impl: Option<&Impl>) -> TokenStream {

--- a/macro/src/lib.rs
+++ b/macro/src/lib.rs
@@ -74,6 +74,7 @@ pub fn bridge(args: TokenStream, input: TokenStream) -> TokenStream {
         .into()
 }
 
+#[doc(hidden)]
 #[proc_macro]
 pub fn type_id(input: TokenStream) -> TokenStream {
     struct TypeId {

--- a/macro/src/lib.rs
+++ b/macro/src/lib.rs
@@ -90,5 +90,5 @@ pub fn type_id(input: TokenStream) -> TokenStream {
     }
 
     let arg = parse_macro_input!(input as TypeId);
-    type_id::expand(arg.path).into()
+    type_id::expand(arg.krate, arg.path).into()
 }

--- a/macro/src/lib.rs
+++ b/macro/src/lib.rs
@@ -37,6 +37,7 @@ use crate::syntax::file::Module;
 use crate::syntax::namespace::Namespace;
 use crate::syntax::qualified::QualifiedName;
 use proc_macro::TokenStream;
+use proc_macro2::TokenTree;
 use syn::parse::{Parse, ParseStream, Parser, Result};
 use syn::parse_macro_input;
 
@@ -75,14 +76,19 @@ pub fn bridge(args: TokenStream, input: TokenStream) -> TokenStream {
 
 #[proc_macro]
 pub fn type_id(input: TokenStream) -> TokenStream {
-    struct TypeId(QualifiedName);
+    struct TypeId {
+        krate: TokenTree,
+        path: QualifiedName,
+    }
 
     impl Parse for TypeId {
         fn parse(input: ParseStream) -> Result<Self> {
-            QualifiedName::parse_quoted_or_unquoted(input).map(TypeId)
+            let krate: TokenTree = input.parse()?;
+            let path = QualifiedName::parse_quoted_or_unquoted(input)?;
+            Ok(TypeId { krate, path })
         }
     }
 
     let arg = parse_macro_input!(input as TypeId);
-    type_id::expand(arg.0).into()
+    type_id::expand(arg.path).into()
 }

--- a/macro/src/lib.rs
+++ b/macro/src/lib.rs
@@ -36,8 +36,8 @@ mod type_id;
 use crate::syntax::file::Module;
 use crate::syntax::namespace::Namespace;
 use crate::syntax::qualified::QualifiedName;
+use crate::type_id::Crate;
 use proc_macro::TokenStream;
-use proc_macro2::TokenTree;
 use syn::parse::{Parse, ParseStream, Parser, Result};
 use syn::parse_macro_input;
 
@@ -77,13 +77,13 @@ pub fn bridge(args: TokenStream, input: TokenStream) -> TokenStream {
 #[proc_macro]
 pub fn type_id(input: TokenStream) -> TokenStream {
     struct TypeId {
-        krate: TokenTree,
+        krate: Crate,
         path: QualifiedName,
     }
 
     impl Parse for TypeId {
         fn parse(input: ParseStream) -> Result<Self> {
-            let krate: TokenTree = input.parse()?;
+            let krate = input.parse().map(Crate::DollarCrate)?;
             let path = QualifiedName::parse_quoted_or_unquoted(input)?;
             Ok(TypeId { krate, path })
         }

--- a/macro/src/type_id.rs
+++ b/macro/src/type_id.rs
@@ -1,14 +1,23 @@
 use crate::syntax::qualified::QualifiedName;
 use proc_macro2::{TokenStream, TokenTree};
-use quote::{format_ident, quote};
+use quote::{format_ident, quote, ToTokens};
 
 pub enum Crate {
     Cxx,
     DollarCrate(TokenTree),
 }
 
+impl ToTokens for Crate {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        match self {
+            Crate::Cxx => tokens.extend(quote!(::cxx)),
+            Crate::DollarCrate(krate) => krate.to_tokens(tokens),
+        }
+    }
+}
+
 // "folly::File" => `(f, o, l, l, y, (), F, i, l, e)`
-pub fn expand(arg: QualifiedName) -> TokenStream {
+pub fn expand(krate: Crate, arg: QualifiedName) -> TokenStream {
     let mut ids = Vec::new();
 
     for word in arg.segments {
@@ -19,11 +28,11 @@ pub fn expand(arg: QualifiedName) -> TokenStream {
             ids.push(match ch {
                 'A'..='Z' | 'a'..='z' => {
                     let t = format_ident!("{}", ch);
-                    quote!(::cxx::#t)
+                    quote!(#krate::#t)
                 }
                 '0'..='9' | '_' => {
                     let t = format_ident!("_{}", ch);
-                    quote!(::cxx::#t)
+                    quote!(#krate::#t)
                 }
                 _ => quote!([(); #ch as _]),
             });

--- a/macro/src/type_id.rs
+++ b/macro/src/type_id.rs
@@ -1,6 +1,11 @@
 use crate::syntax::qualified::QualifiedName;
-use proc_macro2::TokenStream;
+use proc_macro2::{TokenStream, TokenTree};
 use quote::{format_ident, quote};
+
+pub enum Crate {
+    Cxx,
+    DollarCrate(TokenTree),
+}
 
 // "folly::File" => `(f, o, l, l, y, (), F, i, l, e)`
 pub fn expand(arg: QualifiedName) -> TokenStream {

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -413,6 +413,7 @@ mod shared_ptr;
 #[path = "cxx_string.rs"]
 mod string;
 mod symbols;
+mod type_id;
 mod unique_ptr;
 mod unwind;
 #[path = "cxx_vector.rs"]
@@ -428,11 +429,6 @@ pub use crate::unique_ptr::UniquePtr;
 pub use crate::vector::CxxVector;
 pub use crate::weak_ptr::WeakPtr;
 pub use cxxbridge_macro::bridge;
-
-/// For use in impls of the `ExternType` trait. See [`ExternType`].
-///
-/// [`ExternType`]: trait.ExternType.html
-pub use cxxbridge_macro::type_id;
 
 /// Synonym for `CxxString`.
 ///
@@ -466,6 +462,7 @@ pub mod private {
     pub use crate::unwind::catch_unwind;
     pub use crate::vector::VectorElement;
     pub use crate::weak_ptr::WeakPtrTarget;
+    pub use cxxbridge_macro::type_id;
 }
 
 mod actually_private {

--- a/src/type_id.rs
+++ b/src/type_id.rs
@@ -1,0 +1,9 @@
+/// For use in impls of the `ExternType` trait. See [`ExternType`].
+///
+/// [`ExternType`]: trait.ExternType.html
+#[macro_export]
+macro_rules! type_id {
+    ($($path:tt)*) => {
+        $crate::private::type_id! { $($path)* }
+    };
+}

--- a/src/type_id.rs
+++ b/src/type_id.rs
@@ -4,6 +4,6 @@
 #[macro_export]
 macro_rules! type_id {
     ($($path:tt)*) => {
-        $crate::private::type_id! { $($path)* }
+        $crate::private::type_id! { $crate $($path)* }
     };
 }

